### ISSUE TITLE
TLS 1.3 session resumption and ciphers

### ISSUE THREAD

**Issue by graxlop** [proftpd/proftpd#744]
### What I Did

I compiled the latest ProFTPD (git) with the newest OpenSSL 1.1.1 and proftpd doesn't seem to support session resumption when using TLS 1.3. This causes TLS 1.3 capable clients like cURL and Wget to not function properly, since for security reasons they require this to be working.

It also doesn't seem to be using the configured ciphers.

### What I Expected/Wanted

I take it the changes in TLS 1.3 session resumption haven't been implemented yet, since it's quite different from how it worked previously. This is also the case for ciphersuites in TLS 1.3, there are new API functions for this as well.

Some notes on it here:

https://wiki.openssl.org/index.php/TLS1.3#Sessions

**Comment by Castaglia** [comment]
It's true that I've not updated `mod_tls` to support/handle TLS 1.3 yet.

Would you be able to provide the `mod_tls` configuration that you are attempting to use, so that I use that as a starting point for determining what changes are needed?

**Comment by manuelm** [comment]
We recently upgraded OpenSSL and our proftpd services are now supporting TLS 1.3. However soon after we noticed that uploads with our php based webftp service were broken. The file was truncated by a few bytes and the server replied with: `Transfer aborted. Link to file server lost`.

After a few hours of investigation and looking for the cause it turned out that the PHP ftp client implementation only does an unidirectional shutdown. In fact it calls `SSL_shutdown` immediately followed by a `close` (instead of `shutdown`). With TLS 1.2 (and below) this is usually fine. With TLS 1.3 the server sends 2 (per default) sesson tickets after the initial handshake. However these session tickets will never be read as the php ftp client never reads from the data socket. So the hard shutdown in PHP with tickets still in the buffer causes a ECONNRESET on the server side which causes truncation.

Although the hard shutdown in PHP is clearly wrong, I've opted to modify proftpd so it doesn't send any session tickets (`SSL_CTX_set_num_tickets(ssl_ctx, 0);`). Because PHP has 9 `SSL_shutdown` calls sprinkled all over the file which makes fixing it more work.

In contrast this is curls shutdown code which works fine:
https://github.com/curl/curl/blob/e2dd435d473cdc97785df95d032276fafb4b7746/lib/vtls/openssl.c#L1272

References:
https://github.com/openssl/openssl/issues/6904
https://github.com/openssl/openssl/issues/7327

**Comment by Castaglia** [comment]
@manuelm for your use, it sounds like it'd be useful if `mod_tls` had an option for disabling session tickets, yes?  Just for the TLS1.3 case, or all protocol versions?

**Comment by manuelm** [comment]
We've rolled back the patch a while ago as it breaks compatibility with gnutls and the bug in PHP has been fixed (http://git.php.net/?p=php-src.git;a=commit;h=bb4a2e8bb7fc80fa5b3725508bcc8fea525f059b).

There's still an issue regarding truncated files with filezilla (which is using gnutls) + TLS 1.3. I think it's due to the fact the client is closing the socket forcefully which results in an ECONNRESET + truncated files on server side. But that's not session ticket related.
